modules/caddyhttp/caddyauth/caddyauth.go

```diff
@@ -72,8 +77,21 @@ func (a Authentication) ServeHTTP(w http.ResponseWriter, r *http.Request, next c
 	var user User
 	var authed bool
 	var err error
+	// This code makes the following assumption: In case of multiple
+	// authentication providers it is enough if one of them authenticates the
+	// user for the whole authentication to succeed. In case all of the
+	// providers fail to authenticate we write the reply of any of the providers
+	// that redirect to the response writer, if there is no such provider the
+	// response of any provider is written.
```

Comment:
Why do we favor responses with redirects? Why do they get special priority over other responses?

Just want to hash this out, in case it's prudent to have some sort of preference system instead. (I don't mind having implicit simplicity -- as long as it is backed by reason, I just want to understand it. :slightly_smiling_face: )

Reply:
So maybe that is specific to our use case, but the reasoning is as follows; The provider that redirects is the provider that does an OAuth flow and for that flow it is natural that the user get's redirected. E.g. you browse to a page and then you need to login, it redirects you to some identity provider to login, you get back to the page. If we wouldn't favor the redirecting provider, this flow would break.

But I do realize that this might be a bit magic and might even be confusing, e.g. in the case the user actually provider basic auth credentials that are wrong and then gets redirected to the identity provider login page.
Also in case the user agent is not a browser you likely don't want the redirect.

So overall I'm not really sure what's best, I'm also open to suggestions here.

Reply:
Also added the reasoning in the comment.